### DISCUSSION TITLE
OpenPGP Mails

### DISCUSSION THREAD

**Discussion by dhia-gharsallaoui** [wneessen/go-mail#49] [Ideas]
Hello everybody !
thank you for the efforts of making this awesome library.
I'm looking to implement a solution that sends `OpenPGP` encrypted emails.
So i want to ask if it's better to introduce this feature as part of the mail client or not ?
If it's a good idea, I'm happy to contribute.

knowing that `OpenPGP` is no longer supported in the standard golang `crypto` library, but there is a fork where it's maintained. https://github.com/ProtonMail/gopenpgp

thank you in advance.

**Comment by wneessen** [comment]
Hi!

> thank you for the efforts of making this awesome library.
> 
Thank you for using it and being interested in extending it! :) 

> I'm looking to implement a solution that sends `OpenPGP` encrypted emails. 
> So i want to ask if it's better to introduce this feature as part of the mail client or not ? 
> If it's a good idea, I'm happy to contribute.
> 
I feel like OpenPGP should be seperate from the library. Since PGP/GPG only encrypts/signs the mail body, and therefore only
the final/encrypted product would be used with go-mail, I see go-mail more like a consumer than a producer/server for the OpenPGP produced product. Plus I'd like to keep the library only Go standard library dependant, so introducing a 3rd party lib like the ProtonMail client is not really suited.

Still, I'm open for discussion on how a encryption product could be closely integrated with go-mail, i. e. by providing methods that support the product in its functionality.

Hope this makes sense.

**Reply by dhia-gharsallaoui** [reply]
Hi!
You have a reason!
This gives me the idea of implementing a middleware concept in the library:

- Client `WithMiddleware` function that adds a list of middlewares to call them just before sending `FIFO` or `LIFO`.
- `Middleware` is a function that satisfy the take a `Msg` and return a `Msg`.

Then everyone can inject their custom middleware, like my usecase to encrypt the body using openPGP. Others may inject a middleware that takes the mail attachments, uploads them in their blob storage, and changes them with the links. You can imagine many others use cases.

WDYT?
Thank you!

**Reply by wneessen** [reply]
That sounds like a fantastic idea. I really like that!

**Reply by dhia-gharsallaoui** [reply]
Cool !
I will start working on it this Monday. 
I am reachable at the weekend if you have any proposition.
